### ISSUE TITLE
[HOLD for payment 21 June]  iOS - Attachment picker flickers and unable to select for choose document

### ISSUE THREAD

**Issue by isagoico** [Expensify/Expensify.cash#3543]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
Attachment picker displayed and user able to select attachment

## Actual Result:
Attachment picker flickers and unable to choose any document

## Action Performed:
1. Launch the app and login
2. Choose any chat from LHN
3. Tap the + > Add Attachment > Choose Document

## Workaround:
Unknown 

## Platform:
Where is this issue occurring?

 Web
 iOS ✔️ 
 Android
 Desktop App
 Mobile Web

**Version Number:** 1.0.67-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/44479856/121675329-8e486d00-cab3-11eb-8eda-cd51930f48a6.mp4

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cead22 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by Julesssss** [comment]
The [`react-native-web` update ](https://github.com/Expensify/Expensify.cash/pull/3215)seems to have introduced this regression.

**Comment by parasharrajat** [comment]
@Julesssss React-native-web does not affect IOS. It's only active on the web.

**Comment by Julesssss** [comment]
I agree and that makes sense, but perhaps one of the other code changes in that PR introduced the issue?

I can't think of any other explanation for why the issue is not reproducible on `0c8dd5a54`, but is on `e75416eb9`.


https://user-images.githubusercontent.com/10736861/121709122-22283200-cad0-11eb-849e-8d82ae4dde32.MP4

**Comment by parasharrajat** [comment]
I mostly believe it's due to react version upgrade.

**Comment by marcaaron** [comment]
@Julesssss did you happen to test changing the react version to see if it fixed it?

**Comment by Julesssss** [comment]
Nope, but I think that is a likely fix. Perhaps it would be better to remove just that library, but honestly, I think that will take longer as I would need to work out which changes are relevant. Realistically we'll need to revert the full PR.

Also I wasn't clear, but by 'The react-native-web update' I meant the PR, not that specific library.

**Comment by marcaaron** [comment]
>it would be better to remove just that library

Which library?

**Comment by roryabraham** [comment]
I actually think that https://github.com/Expensify/Expensify.cash/pull/3446 is more likely responsible for this issue

**Comment by roryabraham** [comment]
@isagoico do you know what iOS version was used by the tester here?

**Comment by Julesssss** [comment]
@roryabraham I reproduced on iOS 11

> Which library?

` react-native-web`

**Comment by marcaaron** [comment]
Ah no, we can't remove `react-native-web`. Just confirmed document picker works on `main` + `react-native:0.63.4` + simulator

**Comment by roryabraham** [comment]
I just confirmed that the document picker works on `main` + `react-native:0.63.4` + `react-native-document-picker:5.1.0`

**Comment by marcaaron** [comment]
FWIW, it also works fine for me with `0.64.1` 😫

**Comment by roryabraham** [comment]
I was able to reproduce w/ iOS 14, `0.64.1` + `react-native-document-picker:4.0.0`

**Comment by marcaaron** [comment]
testing on iOS 14.5

**Comment by Julesssss** [comment]
Is anyone able to test a physical device? I have a horrible and hopefully incorrect feeling that a previous attachment bug only affected physical devices.

**Comment by roryabraham** [comment]
I can test on a physical iPhone, but not Android.

**Comment by marcaaron** [comment]
Tested Android on my Galaxy S8 works great

**Comment by isagoico** [comment]
Reopening the issue since this is still reproducible on iOS build: 1.0.68-3

https://user-images.githubusercontent.com/44479856/121775966-c79fdc00-cb8a-11eb-9f4a-da2c02c4995f.MP4

**Comment by Julesssss** [comment]
Yep, I can reproduce still on iOS 11, a physical device. I should have clarified on Friday that this was only reproducible on iOS (and possibly only on physical devices).

**Comment by Julesssss** [comment]
Okay, this is only sometimes occurring, but it's a nightmare to verify 😭. In some circumstances it will work the first time you attempt to add an attachment, but then almost all later attempts fail. The only hint is that if you enter a new chat after it fails to display, the Attachment picker will then display -- it's appearing at completely the wrong time, but it remains on screen 😖 

**Works first time**
https://user-images.githubusercontent.com/10736861/121911335-97417480-cd27-11eb-901d-de355d256221.mp4

**Appearing when opening a new chat**

https://user-images.githubusercontent.com/10736861/121913334-374bcd80-cd29-11eb-9763-e95644815a18.mp4

**`main`**
- This works the first time, but any later attempt fails

`main + downgrade 'react-native' to .63.3`
- This seems to have no change at all

**Comment by cead22** [comment]
@Julesssss it looks like you're on this so I'm gonna unassign myself. If you decide to unassign, please let me know and I'll find someone to work on this (which is pretty much all I did Friday)

**Comment by Julesssss** [comment]
I tested builds to locate the introducing commit:
- `5ce0b06a03`: Issue does not occur
- `30d66b717`: Issue does not occur
- `822acf0b56d`:  (main was merged into PR) Issue does not occur
- [`[c0046a2fd]`](https://github.com/Expensify/Expensify.cash/commit/c0046a2fd4f34b208761dcf620d82ce077cb53b7): (main was merged into PR with multiple library changes) Issue is introduced... SO it wasn't the [bump react-native/react-native-web PR](https://github.com/Expensify/Expensify.cash/pull/3215) after all.

There's a high likelihood that one of the following library changes introduced the issue, so next step will be to review them 1 by 1:
- "react-native-onyx" --> Most likely IMO
- "react-native-reanimated": "^2.1.0",
- "react-native-screens": "^3.0.0",
- "react-native": "0.64.1", --> I have ruled this out already
- "@react-native-masked-view/masked-view": "^0.2.4",
- "@react-navigation/drawer": "^6.0.0-next.17"
- "@react-navigation/native": "^6.0.0-next.13"
- "@react-navigation/stack": "^6.0.0-next.25",
- "babel-plugin-transform-remove-console": "^6.9.4",


My main takeaway is that it is frustratingly time-consuming working out which commit/PR introduced a regression. Ideally we would detect more of these regressions at the PR review stage.

**Comment by Julesssss** [comment]
OKAY. Here's the PR which introduced the issue:[ Upgrade react-navigation to 6.x](https://github.com/Expensify/Expensify.cash/pull/3078)

- `37b3591ac`: Issue does not occur
- [`c0766f5`](https://github.com/Expensify/Expensify.cash/commit/c0766f5cf66bb4fb75ef7443c137beb7d8ec97c3): Issue occurs

Specifically, it turns out that bumping `react-native` down to `0.63.3` did solve the problem, which is odd because I saw otherwise when testing that earlier. Now I'll try bumping down the library once more on main to confirm.

**Comment by Julesssss** [comment]
_I hid the above comment as it's not accurate, [this PR](https://github.com/Expensify/Expensify.cash/pull/3446/files) introduced the `react-native` bump._

It gets worse. I can confirm that simply downgrading the 'react-native' library to `0.63.3` on `main` does not fix the issue, [as I discovered](https://github.com/Expensify/Expensify.cash/issues/3543#issuecomment-860754616) earlier today.

So, bumping the library down while on this commit fixes the issue, but not when bumping it down from `main` -- the issue is a combination of two separate changes: 'react-native' AND 'Onyx'... Damn. Reverting Onyx is out of the question as we've already merged a bunch of fixes and would add a lot of extra work. 

We're still blocked by the navigation issue, but I think we should revisit the decision to launch with this issue. Please let me know if you agree/dissagree. Thanks.

CC @trjExpensify

**Comment by parasharrajat** [comment]
@Julesssss I think I found the root cause here. just want you to confirm. 
https://github.com/Expensify/Expensify.cash/blob/bdf2144b6eb1fd561905ea6e3450e8b6997a7a60/src/components/AttachmentPicker/index.native.js#L225-L231

If we increase the Delay here - for eg I set 1 sec, It works fine.

Please check?

**Comment by Julesssss** [comment]
Well done @parasharrajat. Can confirm.

Maybe that's why it sometimes worked correctly the first time? I guess that also explains why two different changes introduced the same bug 🤯 

We had better open up another UpWork contract for you 😅

**Comment by Julesssss** [comment]
I'm assuming you want to take the job? but it would be great if you could raise a PR for this. 

If so we'll get an UpWork task created immediately. Please tag me on the PR so I can review first thing tomorrow 👍

**Comment by parasharrajat** [comment]
Yeah. PR will be up in no time. Upwork can wait.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stephanieelliott (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by Julesssss** [comment]
Hey @stephanieelliott. This one is slightly different as we already have the solution and we just need an UpWork issue created to track payment.

**Comment by stephanieelliott** [comment]
Cool, that's easy! Posted to Upwork, please apply for this one @parasharrajat: https://www.upwork.com/jobs/~01cb0fd992029800cc

**Comment by parasharrajat** [comment]
@Julesssss  PR is up. This deployblocker is done. :cheese:

**Comment by marcaaron** [comment]
@Julesssss so it seems like this isn't related to dependencies at all? Or possibly could be, but not in any obvious way?

**Comment by Julesssss** [comment]
> @Julesssss so it seems like this isn't related to dependencies at all? Or possibly could be, but not in any obvious way?

We've seen this issue before when trying to open a second Modal from another Modal. `onCLose` (or something similar) is called on the first model, but as the second Modal has also started to display, it also is immediately closed. 

I believe both dependencies somehow sped up the displaying of the second Modal, which meant that 10ms was no longer long enough of an artificial delay to wait before showing the second Modal. So yeah, I think both dependencies introduced the problem as a side effect of speeding up the app.
